blockbuster/tob_lane.go

```diff
@@ -0,0 +1,340 @@
+	LaneNameTOB = "tob"
+)
+
+var _ Lane = (*TOBLane)(nil)
+
+type TOBLane struct {
+	logger      log.Logger
+	index       sdkmempool.Mempool
+	af          mempool.AuctionFactory
+	txEncoder   sdk.TxEncoder
+	txDecoder   sdk.TxDecoder
+	anteHandler sdk.AnteHandler
+
+	// txIndex is a map of all transactions in the mempool. It is used
+	// to quickly check if a transaction is already in the mempool.
+	txIndex map[string]struct{}
+}
+
+func NewTOBLane(logger log.Logger, txDecoder sdk.TxDecoder, txEncoder sdk.TxEncoder, maxTx int, af mempool.AuctionFactory, anteHandler sdk.AnteHandler) *TOBLane {
+	return &TOBLane{
+		logger: logger,
+		index: mempool.NewPriorityMempool(
+			mempool.PriorityNonceMempoolConfig[int64]{
+				TxPriority: mempool.NewDefaultTxPriority(),
+				MaxTx:      maxTx,
+			},
+		),
+		af:          af,
+		txEncoder:   txEncoder,
+		txDecoder:   txDecoder,
+		anteHandler: anteHandler,
+		txIndex:     make(map[string]struct{}),
+	}
+}
+
+func (l *TOBLane) Name() string {
+	return LaneNameTOB
+}
+
+func (l *TOBLane) Match(tx sdk.Tx) bool {
+	bidInfo, err := l.af.GetAuctionBidInfo(tx)
+	return bidInfo != nil && err == nil
+}
+
+func (l *TOBLane) Contains(tx sdk.Tx) (bool, error) {
+	txHashStr, err := l.getTxHashStr(tx)
+	if err != nil {
+		return false, fmt.Errorf("failed to get tx hash string: %w", err)
+	}
+
+	_, ok := l.txIndex[txHashStr]
+	return ok, nil
+}
+
+func (l *TOBLane) VerifyTx(ctx sdk.Context, bidTx sdk.Tx) error {
+	bidInfo, err := l.af.GetAuctionBidInfo(bidTx)
+	if err != nil {
+		return fmt.Errorf("failed to get auction bid info: %w", err)
+	}
+
+	// verify the top-level bid transaction
+	ctx, err = l.verifyTx(ctx, bidTx)
+	if err != nil {
+		return fmt.Errorf("invalid bid tx; failed to execute ante handler: %w", err)
+	}
+
+	// verify all of the bundled transactions
+	for _, tx := range bidInfo.Transactions {
+		bundledTx, err := l.af.WrapBundleTransaction(tx)
+		if err != nil {
+			return fmt.Errorf("invalid bid tx; failed to decode bundled tx: %w", err)
+		}
+
+		// bid txs cannot be included in bundled txs
+		bidInfo, _ := l.af.GetAuctionBidInfo(bundledTx)
+		if bidInfo != nil {
+			return fmt.Errorf("invalid bid tx; bundled tx cannot be a bid tx")
+		}
+
+		if ctx, err = l.verifyTx(ctx, bundledTx); err != nil {
+			return fmt.Errorf("invalid bid tx; failed to execute bundled transaction: %w", err)
+		}
+	}
+
+	return nil
+}
+
+// PrepareLane which builds a portion of the block. Inputs a cache of transactions
+// that have already been included by a previous lane.
+func (l *TOBLane) PrepareLane(ctx sdk.Context, maxTxBytes int64, selectedTxs map[string][]byte) ([][]byte, error) {
+	var tmpSelectedTxs [][]byte
+
+	bidTxIterator := l.index.Select(ctx, nil)
+	txsToRemove := make(map[sdk.Tx]struct{}, 0)
+
+	// Attempt to select the highest bid transaction that is valid and whose
+	// bundled transactions are valid.
+selectBidTxLoop:
+	for ; bidTxIterator != nil; bidTxIterator = bidTxIterator.Next() {
```

Comment:
Should we just call `VerifyTx(ctx sdk.Context, bidTx sdk.Tx)` here?

Reply:
Addressed in 851a710